Implement add, sub, mul, div using TensorIterator

colesbury opened the pull request:
```
This adds TensorIterator, a helper class for computing element-wise
operations that's intended to replace the CPU and CUDA apply utils
functions.

CPU kernels are implemented as functions that operate on strided 1-d
tensors compared to CPUApplyUtils which operated individual elements. This
allows the kernels to handle vectorization, while TensorIterator handles
parallelization and non-coalesced dimensions.

GPU kernels continue to operate on elements, but the number of
specializations is reduced. The contiguous case remains the same. The
non-contiguous case uses a single (reduced) shape for all operands and
the fast integer division from THCIntegerDivider. To avoid extra
specializations for indexing with 64-bits, large operations are split
into smaller operations that can be indexed with 32-bits.

Major semantic changes:

 - No more s_add, s_mul, s_div, or s_sub. Broadcasting is handled by
   TensorIterator. The autograd engine performs the reduction assuming
   standard broadcasting if the gradient shape does not match the
   expected shape. Functions that do not use standard broadcasting rules
   should either continue to trace the expand calls or handle the
   reduction in their derivative formula.

 - Use ONNX v7, which supports broadcasting ops.

Performance impact:

 - Small increased fixed overhead (~0.5 us)
 - Larger overhead for wrapped numbers (~2.5 us)
 - No significant change for ops on contiguous tensors
 - Much faster worst-case performance for non-contiguous GPU tensors
 - Faster CPU bias addition (~2x)
 - Faster GPU bias addition (~30% faster)

Future work:

 - Decrease overhead, especially for wrapping numbers in Tensors
 - Handle general inter-type operations
 - Extend to unary ops and reductions
 - Use buffering for compute-bound operations on non-contiguous tensors
   (pull in from CPUApplyUtils)
```

colesbury committed:
Implement add, sub, mul, div using TensorIterator

This adds TensorIterator, a helper class for computing element-wise
operations that's intended to replace the CPU and CUDA apply utils
functions.

CPU kernels are implemented as functions that operate on strided 1-d
tensors compared to CPUApplyUtils which operated individual elements. This
allows the kernels to handle vectorization, while TensorIterator handles
parallelization and non-coalesced dimensions.

GPU kernels continue to operate on elements, but the number of
specializations is reduced. The contiguous case remains the same. The
non-contiguous case uses a single (reduced) shape for all operands and
the fast integer division from THCIntegerDivider. To avoid extra
specializations for indexing with 64-bits, large operations are split
into smaller operations that can be indexed with 32-bits.

Major semantic changes:

 - No more s_add, s_mul, s_div, or s_sub. Broadcasting is handled by
   TensorIterator. The autograd engine performs the reduction assuming
   standard broadcasting if the gradient shape does not match the
   expected shape. Functions that do not use standard broadcasting rules
   should either continue to trace the expand calls or handle the
   reduction in their derivative formula.

 - Use ONNX v7, which supports broadcasting ops.

Performance impact:

 - Small increased fixed overhead (~0.5 us)
 - Larger overhead for wrapped numbers (~2.5 us)
 - No significant change for ops on contiguous tensors
 - Much faster worst-case performance for non-contiguous GPU tensors
 - Faster CPU bias addition (~2x)
 - Faster GPU bias addition (~30% faster)

Future work:

 - Decrease overhead, especially for wrapping numbers in Tensors
 - Handle general inter-type operations
 - Extend to unary ops and reductions
 - Use buffering for compute-bound operations on non-contiguous tensors
   (pull in from CPUApplyUtils)

colesbury committed:
Fixes

 - Use <cstring> instead of <strings.h>
 - Size 1 tensor can be broadcast to size 0 (not the reverse)

colesbury committed:
[TEST] CUDA 8 compatibility fixes

ezyang commented:
CC @houseroad for ONNX changes

ezyang commented on aten/src/ATen/Declarations.cwrap:
```diff
@@ -2141,218 +2141,6 @@
   arguments:
     - THTensor* self
 ]]
-[[
```
Hooray!

ezyang commented on aten/src/ATen/Dispatch_cuda8_compat.h:
```diff
@@ -0,0 +1,76 @@
+
+// Same thing as Dispatch.h but without the anonymous lambdas. This exists to
+// provide compatibility with CUDA 8 which does not allow __device__ lambdas
+// to be defined inside other lambdas.
```
Need a reference here from `Dispatch.h` to remind people to update this file too.

colesbury replied:
I'm going to delete this

colesbury committed:
Clean-up CUDA 8 compat and fix out= resize logic

houseroad commented:
The new ONNX expected files look good to me.

Shall we also remove `fuseExpand` in [peephole](https://github.com/colesbury/pytorch/blob/f2afdf0b54d2b5e7a11c658a1ecbe1f3422c349c/torch/csrc/jit/passes/onnx/peephole.cpp#L81)?

colesbury committed:
Work around GCC 7's flaky Wmaybe-uninitialized

colesbury committed:
Remove __restrict__ (GCC extension)

colesbury committed:
add_out accepts Undefined tensor as result argument

colesbury commented:
@fmassa I didn't see a difference with `__restrict__` in the current code

ezyang commented:
@pytorchbot retest this please

colesbury committed:
Fix handling of empty tensors

colesbury committed:
Attempt to fix Windows CUDA build

colesbury committed:
Attempt to fix Windows build (2)

ezyang commented:
clang under ROCm OOMed (it's earlier in the log so hard to see)

```
ue function 'abs' when argument is of floating point type [-Wabsolute-value]
00:24:09   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:24:09                                                                        ^
00:24:09 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: note: use function 'std::abs' instead
00:24:09   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:24:09                                                                        ^~~~~
00:24:09                                                                        std::abs
00:24:14 1 warning generated.
00:24:16 clang-7.0: error: unable to execute command: Killed
00:24:28 [ 71%] Building HIPCC object caffe2/CMakeFiles/caffe2_hip.dir/__/aten/src/THC/generated/caffe2_hip_generated_THCTensorMathCompareTByte.cu.o
00:24:34 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/THCTensorMathCompareTByte.cu:1:
00:24:34 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/../THCTensorMathCompareT.cuh:8:
00:24:34 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: warning: using integer absolute value function 'abs' when argument is of floating point type [-Wabsolute-value]
00:24:34   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:24:34                                                                        ^
00:24:34 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: note: use function 'std::abs' instead
00:24:34   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:24:34                                                                        ^~~~~
00:24:34                                                                        std::abs
00:24:43 1 warning generated.
00:25:01 LLVM ERROR: out of memory
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN4llvm3sys15PrintStackTraceERNS_11raw_ostreamE+0x2a)[0x16ea92a]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN4llvm3sys17RunSignalHandlersEv+0x3e)[0x16e8a0e]
00:25:01 /opt/rocm/hcc/bin/clang-7.0[0x16e8b5c]
00:25:01 /lib/x86_64-linux-gnu/libpthread.so.0(+0x11390)[0x7fb7837fc390]
00:25:01 /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x38)[0x7fb78256e428]
00:25:01 /lib/x86_64-linux-gnu/libc.so.6(abort+0x16a)[0x7fb78257002a]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN4llvm22report_bad_alloc_errorEPKcb+0x154)[0x1698e54]
00:25:01 /usr/lib/x86_64-linux-gnu/libstdc++.so.6(_Znwm+0x2c)[0x7fb782eafe8c]
00:25:01 /opt/rocm/hcc/bin/clang-7.0[0x120fbb8]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN4llvm11Instruction11setMetadataEjPNS_6MDNodeE+0x23a)[0x12105ea]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN4llvm11Instruction12copyMetadataERKS0_NS_8ArrayRefIjEE+0x58f)[0x11d53df]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZNK4llvm11Instruction5cloneEv+0x5a)[0x11d570a]
00:25:01 /opt/rocm/hcc/bin/clang-7.0[0x171b7e5]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN4llvm25CloneAndPruneIntoFromInstEPNS_8FunctionEPKS0_PKNS_11InstructionERNS_8ValueMapIPKNS_5ValueENS_14WeakTrackingVHENS_14ValueMapConfigISA_NS_3sys10SmartMutexILb0EEEEEEEbRNS_15SmallVectorImplIPNS_10ReturnInstEEEPKcPNS_14ClonedCodeInfoE+0x107)[0x171ffc7]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN4llvm14InlineFunctionENS_8CallSiteERNS_18InlineFunctionInfoEPNS_9AAResultsEbPNS_8FunctionE+0xe91)[0x1749461]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN4llvm17LegacyInlinerBase11inlineCallsERNS_12CallGraphSCCE+0xf4c)[0x129f17c]
00:25:01 /opt/rocm/hcc/bin/clang-7.0[0xd00453]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN4llvm6legacy15PassManagerImpl3runERNS_6ModuleE+0x304)[0x11f8eb4]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN5clang17EmitBackendOutputERNS_17DiagnosticsEngineERKNS_19HeaderSearchOptionsERKNS_14CodeGenOptionsERKNS_13TargetOptionsERKNS_11LangOptionsERKN4llvm10DataLayoutEPNSE_6ModuleENS_13BackendActionESt10unique_ptrINSE_17raw_pwrite_streamESt14default_deleteISM_EEb+0xc17)[0x18c5747]
00:25:01 /opt/rocm/hcc/bin/clang-7.0[0x1fe4d92]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN5clang8ParseASTERNS_4SemaEbb+0x370)[0x27570b0]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN5clang13CodeGenAction13ExecuteActionEv+0x37)[0x1fe4337]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN5clang14FrontendAction7ExecuteEv+0x11e)[0x1c9e86e]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN5clang16CompilerInstance13ExecuteActionERNS_14FrontendActionE+0x146)[0x1c69ae6]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_ZN5clang25ExecuteCompilerInvocationEPNS_16CompilerInstanceE+0x96c)[0x1d3237c]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_Z8cc1_mainN4llvm8ArrayRefIPKcEES2_Pv+0xa18)[0x8df7a8]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(main+0x1951)[0x881871]
00:25:01 /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7fb782559830]
00:25:01 /opt/rocm/hcc/bin/clang-7.0(_start+0x29)[0x8dceb9]
00:25:01 Stack dump:
00:25:01 0.	Program arguments: /opt/rocm/hcc/bin/clang-7.0 -cc1 -D__KALMAR_HC__=1 -D__HCC_HC__=1 -D__KALMAR_CPU__=1 -D__HCC_CPU__=1 -triple x86_64-unknown-linux-gnu -S -disable-free -disable-llvm-verifier -main-file-name THCTensorMode.cu -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -fmath-errno -masm-verbose -mconstructor-aliases -munwind-tables -fuse-init-array -target-cpu x86-64 -dwarf-column-info -debugger-tuning=gdb -coverage-notes-file /var/lib/jenkins/workspace/build/caffe2/CMakeFiles/caffe2_hip.dir/__/aten/src/THC/./caffe2_hip_generated_THCTensorMode.cu.gcno -resource-dir /opt/rocm/hcc/lib/clang/7.0.0 -I/opt/rocm/hcc/bin/../include -I/opt/rocm/hcc/bin/../hcc/include -D __HIPCC__ -I /opt/rocm/hcc/include -I /opt/rocm/hip/include/hip/hcc_detail/cuda -I /opt/rocm/hsa/include -I /opt/rocm/profiler/CXLActivityLogger/include -I /opt/rocm/hip/include -D HIP_VERSION_MAJOR=1 -D HIP_VERSION_MINOR=5 -D HIP_VERSION_PATCH=18234 -D __HIP_ARCH_GFX900__=1 -I /usr/include/x86_64-linux-gnu -I /usr/include/x86_64-linux-gnu/c++/4.2.1 -I /usr/include/c++/4.2.1 -D __HIP_PLATFORM_HCC__=1 -D CUDA_HAS_FP16=1 -D __HIP_NO_HALF_OPERATORS__=1 -D __HIP_NO_HALF_CONVERSIONS__=1 -I /opt/rocm/hip/include -I /opt/rocm/hcc/include -I /opt/rocm/hsa/include -I /opt/rocm/rocrand/include -I /opt/rocm/hiprand/include -I /opt/rocm/rocblas/include -I /opt/rocm/miopen/include -I /data/Thrust -I /data/Thrust/thrust/system/cuda/detail/cub-hip -I -I/opt/rocm/hip/include -I /opt/rocm/hcc/include -I /opt/rocm/hsa/include -I /opt/rocm/rocrand/include -I /opt/rocm/hiprand/include -I /opt/rocm/rocblas/include -I /opt/rocm/miopen/include -I /data/Thrust -I /data/Thrust/thrust/system/cuda/detail/cub-hip -I -I/var/lib/jenkins/workspace/build/caffe2/aten/src/TH -I /var/lib/jenkins/workspace/aten/src/TH -I /var/lib/jenkins/workspace/build/caffe2/aten/src/THC -I /var/lib/jenkins/workspace/aten/src/THC -I /var/lib/jenkins/workspace/aten/src/THCUNN -I /var/lib/jenkins/workspace/aten/src/ATen/cuda -I /var/lib/jenkins/workspace/build/caffe2/aten/src/TH -I /var/lib/jenkins/workspace/aten/src/TH -I /var/lib/jenkins/workspace/aten/src/TH -I /var/lib/jenkins/workspace/aten/src/THC -I /var/lib/jenkins/workspace/build/caffe2/aten/src/TH -I /var/lib/jenkins/workspace/build/caffe2/aten/src/THC -I /var/lib/jenkins/workspace/aten/src -I /var/lib/jenkins/workspace/build/caffe2/aten/src -I /var/lib/jenkins/workspace/build/aten/src -I /var/lib/jenkins/workspace/aten/src/THNN -I /var/lib/jenkins/workspace/aten/src/THCUNN -I /var/lib/jenkins/workspace/aten/src -I /var/lib/jenkins/workspace/aten/../third_party/catch/single_include -I /var/lib/jenkins/workspace/build/caffe2/aten/src/ATen -I /var/lib/jenkins/workspace/aten/src/ATen/.. -I /var/lib/jenkins/workspace/build/caffe2/aten/src/ATen -I /var/lib/jenkins/workspace/build -I /var/lib/jenkins/workspace -I -I/var/lib/jenkins/workspace/third_party/protobuf/src -I /var/lib/jenkins/workspace/cmake/../third_party/eigen -I /var/lib/jenkins/workspace/cmake/../third_party/pybind11/include -I /opt/rocm/hip/include -I /opt/rocm/hipblas/include -I /opt/rocm/hcsparse/include -I /opt/rocm/hcrng/include -I /data/Thrust -I /var/lib/jenkins/workspace/third_party/onnx -I /var/lib/jenkins/workspace/build/third_party/onnx -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/x86_64-linux-gnu/c++/5.4.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/x86_64-linux-gnu/c++/5.4.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/backward -internal-isystem /usr/local/include -internal-isystem /opt/rocm/hcc/lib/clang/7.0.0/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -Wno-deprecated-register -Wno-macro-redefined -Wno-inconsistent-missing-override -Wno-exceptions -Wno-shift-count-negative -Wno-shift-count-overflow -Wno-unused-command-line-argument -std=c++amp -fdeprecated-macro -fdebug-compilation-dir /var/lib/jenkins/workspace/build/caffe2/CMakeFiles/caffe2_hip.dir/__/aten/src/THC -ferror-limit 19 -fmessage-length 0 -fobjc-runtime=gcc -fcxx-exceptions -fexceptions -fdiagnostics-show-option -famp -fhsa-ext -o /tmp/THCTensorMode-618852.s -x hc-host /var/lib/jenkins/workspace/aten/src/THC/THCTensorMode.cu -emit-llvm-bc 
00:25:01 1.	<eof> parser at end of file
00:25:01 2.	Per-module optimization passes
00:25:01 3.	Running pass 'CallGraph Pass Manager' on module '/var/lib/jenkins/workspace/aten/src/THC/THCTensorMode.cu'.
00:25:02 clang-7.0: error: unable to execute command: Aborted (core dumped)
00:25:10 [ 71%] Building HIPCC object caffe2/CMakeFiles/caffe2_hip.dir/__/aten/src/THC/generated/caffe2_hip_generated_THCTensorMathPointwiseByte.cu.o
00:25:16 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/THCTensorMathPointwiseByte.cu:1:
00:25:16 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/../THCTensorMathPointwise.cuh:10:
00:25:16 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: warning: using integer absolute value function 'abs' when argument is of floating point type [-Wabsolute-value]
00:25:16   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:25:16                                                                        ^
00:25:16 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: note: use function 'std::abs' instead
00:25:16   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:25:16                                                                        ^~~~~
00:25:16                                                                        std::abs
00:25:41 [ 71%] Building HIPCC object caffe2/CMakeFiles/caffe2_hip.dir/__/aten/src/THC/generated/caffe2_hip_generated_THCTensorMathCompareByte.cu.o
00:25:47 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/THCTensorMathCompareByte.cu:1:
00:25:47 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/../THCTensorMathCompare.cuh:8:
00:25:47 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: warning: using integer absolute value function 'abs' when argument is of floating point type [-Wabsolute-value]
00:25:47   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:25:47                                                                        ^
00:25:47 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: note: use function 'std::abs' instead
00:25:47   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:25:47                                                                        ^~~~~
00:25:47                                                                        std::abs
00:25:51 1 warning generated.
00:26:00 1 warning generated.
00:26:06 [ 72%] Building HIPCC object caffe2/CMakeFiles/caffe2_hip.dir/__/aten/src/THC/generated/caffe2_hip_generated_THCTensorMathReduceByte.cu.o
00:26:11 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/THCTensorMathReduceByte.cu:1:
00:26:11 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/../THCTensorMathReduce.cuh:6:
00:26:11 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: warning: using integer absolute value function 'abs' when argument is of floating point type [-Wabsolute-value]
00:26:11   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:26:11                                                                        ^
00:26:11 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: note: use function 'std::abs' instead
00:26:11   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:26:11                                                                        ^~~~~
00:26:11                                                                        std::abs
00:26:22 1 warning generated.
00:26:26 clang-7.0: error: clang frontend command failed due to signal (use -v to see invocation)
00:26:26 HCC clang version 7.0.0 (ssh://gerritgit/compute/ec/hcc-tot/clang 86791fc4961dc8ffde77bde20d7dfa5e5cbeff5e) (ssh://gerritgit/compute/ec/hcc-tot/llvm 0ccef158132e1222d549edf2da33d4bc0be6c2d1) (based on HCC 1.2.18184-74f5fa9-86791fc-0ccef15 )
00:26:26 Target: x86_64-unknown-linux-gnu
00:26:26 Thread model: posix
00:26:26 InstalledDir: /opt/rocm/hcc/bin
00:26:26 clang-7.0: note: diagnostic msg: PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
00:26:27 clang-7.0: note: diagnostic msg: 
00:26:27 ********************
00:26:27 
00:26:27 PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
00:26:27 Preprocessed source(s) and associated run script(s) are located at:
00:26:27 clang-7.0: note: diagnostic msg: /tmp/THCTensorMode-34f346.cpp
00:26:27 clang-7.0: note: diagnostic msg: /tmp/THCTensorMode-612e3a.cpp
00:26:27 clang-7.0: note: diagnostic msg: /tmp/THCTensorMode-34f346.sh
00:26:27 clang-7.0: note: diagnostic msg: 
00:26:27 
00:26:27 ********************
00:26:27 [ 72%] Building HIPCC object caffe2/CMakeFiles/caffe2_hip.dir/__/aten/src/THC/generated/caffe2_hip_generated_THCTensorMaskedByte.cu.o
00:26:32 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/THCTensorMaskedByte.cu:1:
00:26:32 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/../THCTensorMasked.cuh:7:
00:26:32 In file included from /var/lib/jenkins/workspace/aten/src/THC/THCReduce.cuh:13:
00:26:32 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: warning: using integer absolute value function 'abs' when argument is of floating point type [-Wabsolute-value]
00:26:32   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:26:32                                                                        ^
00:26:32 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: note: use function 'std::abs' instead
00:26:32   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:26:32                                                                        ^~~~~
00:26:32                                                                        std::abs
00:26:39 1 warning generated.
00:26:45 [ 72%] Building HIPCC object caffe2/CMakeFiles/caffe2_hip.dir/__/aten/src/THC/generated/caffe2_hip_generated_THCTensorSortChar.cu.o
00:26:51 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/THCTensorSortChar.cu:1:
00:26:51 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/../THCTensorSort.cuh:5:
00:26:51 In file included from /var/lib/jenkins/workspace/aten/src/THC/THCSortUtils.cuh:6:
00:26:51 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: warning: using integer absolute value function 'abs' when argument is of floating point type [-Wabsolute-value]
00:26:51   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:26:51                                                                        ^
00:26:51 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: note: use function 'std::abs' instead
00:26:51   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:26:51                                                                        ^~~~~
00:26:51                                                                        std::abs
00:26:55 1 warning generated.
00:26:58 [ 72%] Building HIPCC object caffe2/CMakeFiles/caffe2_hip.dir/__/aten/src/THC/generated/caffe2_hip_generated_THCTensorMathCompareTChar.cu.o
00:27:04 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/THCTensorMathCompareTChar.cu:1:
00:27:04 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/../THCTensorMathCompareT.cuh:8:
00:27:04 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: warning: using integer absolute value function 'abs' when argument is of floating point type [-Wabsolute-value]
00:27:04   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:27:04                                                                        ^
00:27:04 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: note: use function 'std::abs' instead
00:27:04   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:27:04                                                                        ^~~~~
00:27:04                                                                        std::abs
00:27:07 [ 72%] Building HIPCC object caffe2/CMakeFiles/caffe2_hip.dir/__/aten/src/THC/generated/caffe2_hip_generated_THCTensorMathPointwiseChar.cu.o
00:27:12 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/THCTensorMathPointwiseChar.cu:1:
00:27:12 In file included from /var/lib/jenkins/workspace/aten/src/THC/generated/../THCTensorMathPointwise.cuh:10:
00:27:12 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: warning: using integer absolute value function 'abs' when argument is of floating point type [-Wabsolute-value]
00:27:12   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:27:12                                                                        ^
00:27:12 /var/lib/jenkins/workspace/aten/src/THC/THCNumerics.cuh:740:72: note: use function 'std::abs' instead
00:27:12   static inline __host__ __device__  double abs  (double a) { return   ::abs(a); }
00:27:12                                                                        ^~~~~
00:27:12                                                                        std::abs
00:27:23 1 warning generated.
00:27:24 clang-7.0: error: clang frontend command failed due to signal (use -v to see invocation)
00:27:24 HCC clang version 7.0.0 (ssh://gerritgit/compute/ec/hcc-tot/clang 86791fc4961dc8ffde77bde20d7dfa5e5cbeff5e) (ssh://gerritgit/compute/ec/hcc-tot/llvm 0ccef158132e1222d549edf2da33d4bc0be6c2d1) (based on HCC 1.2.18184-74f5fa9-86791fc-0ccef15 )
00:27:24 Target: x86_64-unknown-linux-gnu
00:27:24 Thread model: posix
00:27:24 InstalledDir: /opt/rocm/hcc/bin
00:27:24 clang-7.0: note: diagnostic msg: PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
00:27:25 clang-7.0: note: diagnostic msg: 
00:27:25 ********************
00:27:25 
00:27:25 PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
00:27:25 Preprocessed source(s) and associated run script(s) are located at:
00:27:25 clang-7.0: note: diagnostic msg: /tmp/THCTensorIndex-5c2251.cpp
00:27:25 clang-7.0: note: diagnostic msg: /tmp/THCTensorIndex-431cfa.cpp
00:27:25 clang-7.0: note: diagnostic msg: /tmp/THCTensorIndex-5c2251.sh
00:27:25 clang-7.0: note: diagnostic msg: 
00:27:25 
00:27:25 ********************
```

ezyang commented:
CC @Jorghi12 @bddppq

colesbury committed:
Remove is_test from Dropout

colesbury committed:
Fixes for ROCm

colesbury committed:
Attempt to please both CUDA and HIP

colesbury committed:
Replace __CUDACC__ with __NVCC__

colesbury committed:
Hack to work around incorrect shape in ONNX RNN export

ezyang commented:
SHIP IT SHIP IT

ezyang commented on aten/src/ATen/ExpandUtils.cpp:
```diff
@@ -21,7 +21,7 @@ std::vector<int64_t> infer_size(IntList a, IntList b) {
         ") at non-singleton dimension ", i);
 
-    expandedSizes[i] = std::max(sizeA, sizeB);
+    expandedSizes[i] = sizeA == 1 ? sizeB : sizeA;
```
Bugfix?

gchanan replied:
Yes (I think, the old behavior may have been there to preserve TH-style semantics, although I don't remember off the top of my head if that was intentional or not).  Anyway, I have the same fix in one of my n-dimensional 0-size branches and tests.

ezyang commented on aten/src/ATen/ExpandUtils.h:
```diff
@@ -130,4 +130,39 @@ inline std::vector<Tensor> expand_outplace(TensorList to_expand) {
 
+// Sums `tensor` repeatedly to produce a tensor of shape `shape`.
+// Precondition: is_expandable_to(shape, tensor.sizes()) must be true
+static inline Tensor reduce_to(Tensor tensor, IntList shape) {
```
If it sums, shouldn't it be called `sum_to`?

apaszke commented on aten/src/ATen/cuda/detail/OffsetCalculator.cuh:
```diff
@@ -0,0 +1,63 @@
+
+template <int NARGS>
+struct OffsetCalculator {
+  static const int MAX_DIMS = 25;
```
`constexpr`?

li-roy commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -0,0 +1,168 @@
+#include <ATen/Dispatch.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/native/TensorIterator.h>
+#include <time.h>
```
remove this?

li-roy commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -0,0 +1,514 @@
+
+#include <ATen/ExpandUtils.h>
+#include <ATen/Parallel.h>
+#include <time.h>
```
remove?

colesbury commented on aten/src/ATen/TensorImpl.h:
```diff
@@ -65,6 +65,18 @@ struct TensorImpl : public Retainable {
   }
 
+  // True if a tensor was auto-wrapped from a C++ or Python number.
+  // Wrapped numbers are considered "lower priority" when computing the result
```
Wrapped numbers do not participate in the result type calculation if there are any Tensors that are not wrapped numbers. (I'll improve the comment).

This is intended to keep backwards compatibility. We wrap all integers as 0-d LongTensors and Python floats as DoubleTensors, but we don't want to promote `byte_tensor + 5` to long or `float_tensor + 3.5` to double.

NumPy result_type calculation is a bit different. I'm not sure we want to use it because it promotes float arrays to double arrays in more cases than we would like.

apaszke replied:
Makes sense. Still, we should take them into account when you do `byte_tensor + 2.5`, right?

colesbury commented on aten/src/ATen/cuda/detail/OffsetCalculator.cuh:
```diff
@@ -0,0 +1,63 @@
+  static const int MAX_DIMS = 25;
+
+  // The offset for each argument (in bytes). Wrapper around fixed-size array.
+  struct offsets_t {
```
You can't use arrays as return types in C++. On the host you could use `std::array<uint32_t, NARGS>`, but CUDA support for `std::array` isn't great.

gchanan commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -0,0 +1,514 @@
+  // 3) Any tensor
+  //
+  // For example:
+  //  torch.randn(3, 3) + torch.tensor([1.0], dtype=torch.double) -> torch.double
```
the LHS here and below don't specify a dtype (it depends on what the default is set to).

gchanan commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -0,0 +1,514 @@
+  // The result dtype is computed with the precedence:
+  // 1) Tensors of rank one or higher
+  // 2) Tensors of any rank that aren't wrapped numbers
+  // 3) Any tensor
```
Aren't these just wrapped numbers, then?

gchanan commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -0,0 +1,514 @@
+  auto result_type = ScalarType::Undefined;
+  auto backend = Backend::Undefined;
+  std::tie(result_type, backend) = compute_result_type(operands_, [](const Tensor& t) {
+    return t.dim() > 0;
```
interestingly this can change the result type of operations when we turn on n-dimensional empty tensors.

gchanan commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -0,0 +1,514 @@
+  for (auto& op : operands_) {
+    if (!op.type_) {
+      op.type_ = &type;
+      if (op.tensor_->defined() && type != op.tensor_->type()) {
```
we pass undefined tensors to this?

gchanan commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -0,0 +1,514 @@
+}
+
+DimVector TensorIterator::invert_perm(IntList input) const {
+  // Invert the permutation caused by reorder_dimensions. This is not valid
```
might want to just record that coalesce_dimensions was called and assert it wasn't called here.

gchanan commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -0,0 +1,514 @@
+    return true;
+  };
+
+  auto copy_strides = [&](int dim0, int dim1) {
```
nit: I'd call this coalesec_stride_dim or something -- 'copying' makes me think of a different op.

gchanan commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -0,0 +1,514 @@
+  AT_ASSERT(dim < ndim() && size >= 1);
+  shape_[dim] = size;
+  for (auto& op : operands_) {
+    op.data_ = ((char*)op.data_) + op.stride_[dim] * start;
```
do we check for empty tensors somewhere?  This math may or may not be correct for n-dimensional empty tensors.

gchanan commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -0,0 +1,514 @@
+
+    auto shape = op.tensor_->sizes();
+    if (shape_.empty()) {
+      shape_ = shape;
```
why do we need to do this?  doesn't infer_size handle this?  Or maybe it will after https://github.com/pytorch/pytorch/pull/9209?

gchanan commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -0,0 +1,514 @@
+  }
+
+  // Outputs cannot be broadcasted. Check that the shape of the outputs matches
+  // the inferred shape. There's
```
missing rest of comment?

gchanan commented on aten/src/ATen/native/cpu/Loops.h:
```diff
@@ -0,0 +1,137 @@
+using namespace vec256;
+
+// all three operands contiguous
+static inline bool is_binary_contiguous(const int64_t* strides, const int64_t* element_sizes) {
```
are these used?  I don't see it, but maybe github is hiding the usages.

gchanan commented on aten/src/ATen/native/cpu/Loops.h:
```diff
@@ -0,0 +1,137 @@
+
+// Basic loop binary operation (two inputs, one output). May be auto-vectorized
+// by the compiler.
+#define BASE_BINARY_LOOP(op) \
```
are these all safe for n-dimensional empty tensors?

colesbury committed:
Changes from review

colesbury commented:
@pytorchbot retest this please

gchanan commented on aten/src/ATen/native/BinaryOps.h:
```diff
@@ -7,11 +7,11 @@ namespace at { struct TensorIterator; }
 namespace at { namespace native {
 
-using add_fn = void(*)(TensorIterator&, Scalar alpha);
+using binary_fn_alpha = void(*)(TensorIterator&, Scalar alpha);
```
much better!

gchanan commented on aten/src/ATen/TensorImpl.h:
```diff
@@ -66,9 +66,10 @@ struct TensorImpl : public Retainable {
+  // Wrapped numbers do not participate in the result type computation for
+  // mixed- type operations if there are any Tensors that are not wrapped
+  // numbers. In other respsects, they behave like their non-wrapped, zero-dim
+  // equivalents. See TensorIterator.h for result type computation.
```
prefer: See [Result type computation] in TensorIterator.h

gchanan commented on aten/src/ATen/native/TensorIterator.h:
```diff
@@ -9,6 +9,8 @@
 // arithmetic, comparisions, and trigonometric functions. It handles
 // broadcasting and type conversions of operands.
 //
+// This inspired by NumPy's Array Iterator API (NpyIter).
```
This is.

gchanan commented on aten/src/ATen/native/TensorIterator.h:
```diff
@@ -20,49 +22,61 @@
+// Note [Result type computation]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// TensorIterator handles limited mixed-type operations. The result type is
+// computed using promoteTypes on the types of the operands with the following
```
I'd try to avoid saying types in ATen ("on the types)"...you mean scalar types here, correct?

gchanan commented on aten/src/ATen/native/TensorIterator.h:
```diff
@@ -20,49 +22,61 @@
+//
+// So if there are any tensors of dim 1 or higher, then 0-dim tensors do not
+// affect the result type. This behavior was chosen to preserve backwards
+// compatibility and is *likely to change* in the near future.
```
Why is this likely to change?

colesbury replied:
Currently, `int_tensor + 2.5` returns an int tensor. This behavior is preserved in this PR. In the future, if we adopt https://github.com/pytorch/pytorch/issues/9515, it will return a floating-point tensor.

gchanan commented on aten/src/ATen/native/TensorIterator.h:
```diff
@@ -193,11 +216,14 @@ struct SplitUntil32Bit {
+    bool operator!=(const iterator& other) const { return !(*this == other); }
 
-    /// stack of  queue
+    /// stack of TensorIterator's to be split
```
nit: TensorIterators or TensorIterator(s).

gchanan commented on aten/src/ATen/native/cpu/Loops.h:
```diff
@@ -10,126 +10,145 @@ namespace at { namespace native { namespace {
-         strides[1] == element_sizes[1] &&
-         strides[2] == element_sizes[2];
+template <typename traits>
+static inline bool is_binary_contiguous(const int64_t* strides) {
```
does it slow things down to use an IntList here?

colesbury committed:
Fix comments from review

colesbury committed:
Handle `1 - tensor`

torch.sub allows either self or other to be scalars.

colesbury committed:
Changes for dispatch_stub_v2

colesbury committed:
Fixes after rebase

ezyang commented:
@pytorchbot retest this please

colesbury commented:
@pytorchbot retest this please

colesbury commented:
@pytorchbot retest this please

colesbury commented:
@pytorchbot retest this please

apaszke commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -273,11 +276,12 @@ void PropagateShapeOnNode(Node * node, bool insert_expands) {
+      node->matches("aten::mul(Tensor self, Tensor other) -> Tensor")) {
+    PropagateShapeOnNodeByRunningIt(node);
+    return;
+  } else if (insert_expands && (
```
Why is only such a small subset of nodes excluded from the `insert_expands` logic?

colesbury replied:
Because these are the only nodes that handle expands internally

apaszke commented on test/expect/TestJit.test_fuse_last_device.expect:
```diff
@@ -1,16 +1,21 @@
-  %6 : Float(1) = aten::add(%9, %5, %5)
-  %3 : Float(1) = aten::tanh(%6)
+with prim::FusionGroup_1 = graph(%2 : Float(1)) {
+  %3 : Float(1) = aten::tanh(%2)
```
This test is broken

colesbury committed:
Add comment and temporarily change test_fuse_last_device

Fusion of constant addition no longer works and needs to be fixed.
This changes test_fuse_last_device to avoid it.

colesbury committed:
Remove code superseded by @apaszke's recent PR
